serverkeyapi/internal/api.go

```diff
@@ -45,51 +79,153 @@ func (s *ServerKeyAPI) FetchKeys(
-	// First consult our local database and see if we have the requested
+
+	// First, check if any of these key checks are for our own keys. If
+	// they are then we will satisfy them directly.
+	for req := range requests {
+		if req.ServerName == s.Cfg.Matrix.ServerName {
+			// We found a key request that is supposed to be for our own
+			// keys. Remove it from the request list so we don't hit the
+			// database or the fetchers for it.
+			delete(requests, req)
+
+			// Look up our own keys.
+			request := &api.QueryLocalKeysRequest{}
+			response := &api.QueryLocalKeysResponse{}
+			if err := s.QueryLocalKeys(ctx, request, response); err != nil {
+				return nil, err
+			}
+
+			// Depending on whether the key is expired or not, we'll need
+			// to write slightly different
+			if verifyKeys, ok := response.ServerKeys.VerifyKeys[req.KeyID]; ok {
+				// The key is current.
+				results[req] = gomatrixserverlib.PublicKeyLookupResult{
+					VerifyKey:    verifyKeys,
+					ExpiredTS:    gomatrixserverlib.PublicKeyNotExpired,
+					ValidUntilTS: response.ServerKeys.ValidUntilTS,
+				}
+			} else if verifyKeys, ok := response.ServerKeys.OldVerifyKeys[req.KeyID]; ok {
+				// The key is expired.
+				results[req] = gomatrixserverlib.PublicKeyLookupResult{
+					VerifyKey:    verifyKeys.VerifyKey,
+					ExpiredTS:    verifyKeys.ExpiredTS,
+					ValidUntilTS: gomatrixserverlib.PublicKeyNotValid,
+				}
+			}
+		}
+	}
+
+	// Then consult our local database and see if we have the requested
 	// keys. These might come from a cache, depending on the database
 	// implementation used.
 	if dbResults, err := s.OurKeyRing.KeyDatabase.FetchKeys(ctx, requests); err == nil {
-		// We successfully got some keys. Add them to the results and
-		// remove them from the request list.
+		// We successfully got some keys. Add them to the results.
 		for req, res := range dbResults {
-			if !res.WasValidAt(now, true) {
-				// We appear to be past the key validity. Don't return this
-				// key with the results.
-				continue
-			}
 			results[req] = res
-			delete(requests, req)
+
+			// If the key is valid right now then we can also remove it
+			// from the request list as we don't need to fetch it again
+			// in that case.
+			if res.WasValidAt(now, true) {
+				delete(requests, req)
+			}
 		}
 	}
+
 	// For any key requests that we still have outstanding, next try to
 	// fetch them directly. We'll go through each of the key fetchers to
-	// ask for the remaining keys.
+	// ask for the remaining keys
 	for _, fetcher := range s.OurKeyRing.KeyFetchers {
+		// If there are no more keys to look up then stop.
 		if len(requests) == 0 {
 			break
 		}
-		if fetcherResults, err := fetcher.FetchKeys(ctx, requests); err == nil {
-			// We successfully got some keys. Add them to the results and
-			// remove them from the request list.
+		fetcherCtx, fetcherCancel := context.WithTimeout(ctx, time.Second*30)
+		defer fetcherCancel()
```

Comment:
This won't behave as you expect. Put the body of the `for` into a separate function if you wish to defer cancel.

Reply:
I've rearranged this all now, but I'm aware that the `defer` wouldn't have ran until the function ended. It was mostly to avoid leaky contexts.